rs/embedders/benches/heap.rs

```diff
@@ -72,57 +90,69 @@ fn throughput(size: Size, step: Step) -> Option<Throughput> {
 fn setup_action(src: Src) -> SetupAction {
     match src {
         Src::Checkpoint => SetupAction::PerformCheckpoint,
+        Src::Mix => SetupAction::PerformCheckpointCallSetup,
         Src::PageDelta | Src::NewAllocation => SetupAction::None,
     }
 }
 
-fn heap_op(op: Op) -> String {
-    match op {
-        Op::Read => "(drop (i64.load (local.get $address)))".into(),
-        Op::Write => "(i64.store (local.get $address) (global.get $counter))".into(),
+fn heap_op(op: Op, mem: Mem) -> String {
+    match (op, mem) {
+        (Op::Read, Mem::Wasm32) => {
+            "(global.set $data (i64.load8_u (i32.wrap_i64 (local.get $address))))"
+        }
+        (Op::Read, Mem::Wasm64) => "(global.set $data (i64.load8_u (local.get $address)))",
+        (Op::Write, Mem::Wasm32) => {
+            "(i64.store8 (i32.wrap_i64 (local.get $address)) (global.get $counter))"
+        }
+        (Op::Write, Mem::Wasm64) => "(i64.store8 (local.get $address) (global.get $counter))",
     }
+    .into()
 }
 
 fn heap_func_init_body(mem: Mem, size: Size, src: Src) -> String {
     let wasm_pages = (size as usize).div_ceil(WASM_PAGE_SIZE_IN_BYTES);
     match src {
-        Src::Checkpoint | Src::PageDelta => String::new(),
+        Src::Checkpoint | Src::PageDelta | Src::Mix => String::new(),
         Src::NewAllocation => format!("(drop (memory.grow (i{mem}.const {wasm_pages})))"),
     }
 }
 
-fn loop_body(op: &str, mem: Mem, dir: Dir, size: Size, step: Step) -> String {
-    let size = size as usize;
+fn loop_body(op: &str, dir: Dir, initial_offset: usize, size: Size, step: Step) -> String {
     let step = step as usize;
+    let end = size as usize - 1;
     match dir {
-        Dir::Forward => format!(
-            r#"
-                (local.set $address (i{mem}.const 0))
+        Dir::Fwd => {
+            format!(
+                r#"
+                (local.set $address (i64.const {initial_offset}))
```

Comment:
What does the initial offset accomplish?

Reply:
We need the offset to jump over the odd pages, i.e. 1, 3, 5, etc. I see the issue. I believe I've fixed it.